src/Features/LanguageServer/Protocol/CustomProtocol/FindUsagesLSPContext.cs

```diff
@@ -0,0 +1,291 @@
+            if (text == null)
+            {
+                return null;
+            }
+
+            // TO-DO: The Origin property should be added once Rich-Nav is completed.
+            // https://github.com/dotnet/roslyn/issues/42847
+            var result = new LSP.VSReferenceItem
+            {
+                ContainingMember = properties.TryGetValue(
+                    AbstractReferenceFinder.ContainingMemberInfoPropertyName, out var referenceContainingMember) ? referenceContainingMember : null,
+                ContainingType = properties.TryGetValue(
+                    AbstractReferenceFinder.ContainingTypeInfoPropertyName, out var referenceContainingType) ? referenceContainingType : null,
+                DefinitionId = definitionId,
+                DefinitionText = definitionText,    // Only definitions should have a non-null DefinitionText
+                DisplayPath = location?.Uri.LocalPath,
+                DocumentName = documentSpan == default ? null : documentSpan.Document.Name,
+                Id = id,
+                Kind = symbolUsageInfo.HasValue ? ProtocolConversions.SymbolUsageInfoToReferenceKinds(symbolUsageInfo.Value) : new ReferenceKind[] { },
+                Location = location,
+                ProjectName = documentSpan == default ? null : documentSpan.Document.Project.Name,
+                ResolutionStatus = ResolutionStatusKind.ConfirmedAsReference,
+                Text = text,
+            };
+
+            return result;
+
+            static async Task<LSP.Location?> ComputeLocationAsync(
+                Document document,
+                int position,
+                DocumentSpan documentSpan,
+                IMetadataAsSourceFileService metadataAsSourceFileService,
+                CancellationToken cancellationToken)
+            {
+                // If we have no source span, our location may be in metadata.
+                if (documentSpan == default)
+                {
+                    var symbol = await SymbolFinder.FindSymbolAtPositionAsync(document, position, cancellationToken).ConfigureAwait(false);
+
+                    if (symbol == null || symbol.Locations == null || symbol.Locations.IsEmpty || !symbol.Locations.First().IsInMetadata)
+                    {
+                        // We couldn't find the location in metadata and it's not in any of our known documents.
+                        return null;
+                    }
+
+                    var declarationFile = await metadataAsSourceFileService.GetGeneratedFileAsync(
+                        document.Project, symbol, allowDecompilation: false, cancellationToken).ConfigureAwait(false);
+
+                    var linePosSpan = declarationFile.IdentifierLocation.GetLineSpan().Span;
+
+                    return new LSP.Location
+                    {
+                        Uri = new Uri(declarationFile.FilePath),
```

Comment:
@dibarbet @jasonmalinowski  can this ever fail?  are all filepaths safely convertible to URIs?

Reply:
With razor, not necessarily which was where [this ](http://sourceroslyn.io/#Microsoft.CodeAnalysis.LanguageServer.Protocol/Extensions/ProtocolConversions.cs,3d4591e837306903)came from...

Otherwise I think file paths in general are all convertible to URIs (with file:///), which should be handled by the URI constructor. See [remarks](https://docs.microsoft.com/en-us/dotnet/api/system.uri?view=netframework-4.8#remarks).
Whether the file path specified in the declarationfile string is convertible, I can't say - the constructor can [throw](https://docs.microsoft.com/en-us/dotnet/api/system.uri.-ctor?view=netframework-4.8#System_Uri__ctor_System_String_).